docs/design-proposals/2022-04 TraceQL Concepts.md

```diff
@@ -0,0 +1,198 @@
+- [Structure](#structure)
+- [Selecting spans](#selecting-spans)
+  - [Intrinsic fields](#intrinsic-fields)
+  - [Attribute fields](#attribute-fields)
+  - [Field expressions](#field-expressions)
+- [Combining Spansets](#combining-spansets)
+  - [Structural Operators](#structural-operators)
+- [Aggregators](#aggregators)
+- [Pipelining](#pipelining)
+- [Grouping](#grouping)
+- [Examples](#examples)
+  - [Shorthand](#shorthand)
+
+## Capabilities
+
+A TraceQL query can select traces based on 
+
+- span attributes, timing and duration
+- structural relationships between spans
+- aggregated data from the spans in a trace
+
+The TraceQL language is designed to use the same syntax and semantics as PromQL and LogQL where possible. The syntax and semantics differ due to the specialized needs of querying traces.
+
+## Structure
+
+A query is an expression that is evaluated on one trace at a time, and iterates, evaluating each trace. The expression is evaluated on each trace as a whole, one at a time. If the evaluation produces a set of spans, then the trace and the set of spans is part of the result set.
+
+The structure of a query is an expression or set of chained expressions (a pipeline). Chained expressions are evaluated left to right.  Each expression selects or discards this trace from being included in the results set.
+
+## Selecting spans
+
+In TraceQL the curly brackets `{}` always select a set of spans from the current trace. They will commonly be paired with a condition to reduce the spans being passed in.
+
+`{ http.status = 200 }`
+
+This simple query will be evaluated on every span of every trace, one at a time. 
+
+If the trace being evaluated contains *no* spans with an attribute`http.status` with the value `200` then no spans will be selected and this trace will not appear in our resultset.
+If the trace does contain spans with an attribute`http.status` with the value `200` then *only* those spans will be returned. The trace is reduced to only the set of spans that match the condition inside the `{}`. This set of spans (and therefore trace) will be included in our resultset.
+
+### Intrinsic fields
+
+Every span has certain intrinsic fields that can be referenced when selecting spans
+
+| field name | description |
+| ---------- | ----------- |
+| `start`    | start time of the span |
+| `end`      | end time of the span |
+| `duration` | end - start time of the span |
+| `name`     | operation or span name |
+| `status`   | status values are "error", "ok", or "unset" |
+| `parent`   | the parent of this span | 
+
+**Examples**
+Find traces that contain spans whose duration is greater than 2 seconds:  
+`{ duration > 2s }`
+
+Find traces that contain a span named "HTTP POST":  
+`{ name = "HTTP POST"  }`
+
+Find traces that cross service boundaries:  
+`{ parent.service.name != service.name }`
+
+### Attribute fields
+
+We can refer to dynamic attributes (also known as tags) on the span or the span's resource.
```

Comment:
Does this imply that there is no way to distinguish between resource attributes and span attributes?

Reply:
In our internal doc we have suggested using a prefix like the below to explicitly select resource or span attributes. If "span" or "resource" is not present then we would look on both the resource and the span. The exact details here are still under  discussion. Input is appreciated.

`{ span.key = "val" }`
`{ resource.key = "val" }`

I kind of hate the term "resource" and prefer the Jaeger terminology "process", but we will likely align with OTel on naming conventions.

Reply:
> If "span" or "resource" is not present then we would look on both the resource and the span

I would love the ability to specify span-level or resource-level, if no reason other than improved search times and efficiency.  From my standpoint, I think we're waiting for the right syntax.   `{ key=val }` is great for usability but not precise... And ` {span.key=val}` feels wordy, and could conflict if another attribute is named `span.key`.

Reply:
Note: there's a new OTEP for a new "scope attribute".